File: java/Algorithms/src/Collections/ResizingArrayQueue.java
package Collections;

import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * Created by jianming.xiao on 10/8/14.
 */
public class ResizingArrayQueue<Item> implements IQueue<Item> {
    private Item[] q;
    private int N = 0;
    private int first = 0;
    private int last = 0;

    /**
     * Initializes an empty queue.
     */
    public ResizingArrayQueue() {
        this.q = (Item[]) new Object[2];
    }

    /**
     * Is this queue empty?
     *
     * @return true if this queue is empty; false otherwise
     */
    @Override
    public boolean isEmpty() {
        return N == 0;
    }

    /**
     * Returns the number of items in this queue.
     *
     * @return the number of items in this queue
     */
    @Override
    public int size() {
        return N;
    }

    private void resize(int capacity) {
        Item[] temp = (Item[]) new Object[capacity];
        for (int i = 0; i < N; i++) {
            temp[i] = q[(first + i) % q.length];
        }
        q = temp;
        first = 0;
        last = N;
    }

    /**
     * Adds the item to this queue.
     *
     * @param item the item to add
     * @throws NullPointerException if the <tt>item</tt> is null;
     */
    @Override
    public void enqueue(Item item) {
        if (item == null) throw new NullPointerException();
        if (N == q.length) resize(q.length * 2);
        q[last++] = item;
        if (last == q.length) last = 0;
        N++;
    }

    /**
     * Removes and returns the item on this queue that was least recently added.
     *
     * @return the item on this queue that was least recently added
     * @throws java.util.NoSuchElementException if this queue is empty
     */
    @Override
    public Item dequeue() {
        if (isEmpty()) throw new NoSuchElementException();
        Item result = q[first];
        q[first] = null;
        N--;
        first++;
        if (first == q.length) first = 0;
        if (N == q.length / 4) resize(q.length / 2);
        return result;
    }

    /**
     * Returns the item least recently added to this queue.
     *
     * @return the item least recently added to this queue
     * @throws java.util.NoSuchElementException if this queue is empty
     */
    @Override
    public Item peek() {
        if (isEmpty()) throw new NoSuchElementException();
        return q[first];
    }

    /**
     * Returns a string representation of this queue.
     *
     * @return the sequence of items in FIFO order, separated by spaces
     */
    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder();
        for (Item item : this) {
            stringBuilder.append(item.toString()).append(" ");
        }
        return stringBuilder.toString();
    }

    /**
     * Returns an iterator that iterates over the items in this queue in FIFO order.
     *
     * @return an iterator that iterates over the items in this queue in FIFO order
     */
    @Override
    public Iterator<Item> iterator() {
        return new ResizingArrayQueueIterator();
    }

    private class ResizingArrayQueueIterator implements Iterator<Item> {
        private int i = 0;

        @Override
        public boolean hasNext() {
            return i < N;
        }

        @Override
        public Item next() {
            if (!hasNext()) throw new NoSuchElementException();
            return q[(first + i) % q.length];
        }

        @Override
        public void remove() {
            throw new UnsupportedOperationException();
        }
    }
}

File: c#/BigEgg.Algorithm/BigEgg.Algorithm/Collections/ResizingArrayQueue.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace BigEgg.Algorithm.Collections
{
    public class ResizingArrayQueue<T> : IQueue<T>
    {
        private int N;
        private int first;
        private int last;
        private T[] q;

        /// <summary>
        /// Initializes a new instance of the <see cref="ResizingArrayQueue{T}"/> class.
        /// </summary>
        public ResizingArrayQueue()
        {
            q = new T[2];
            N = 0;
            first = 0;
            last = 0;
        }

        /// <summary>
        /// Is this queue empty?
        /// </summary>
        /// <returns><c>True</c> if this queue is empty; <c>false</c> otherwise</returns>
        public bool IsEmpty()
        {
            return N == 0;
        }

        /// <summary>
        /// Returns the number of items in this queue.
        /// </summary>
        /// <returns>The number of items in this queue.</returns>
        public int Size()
        {
            return N;
        }

        /// <summary>
        /// Resize the underlying array holding the elements.
        /// </summary>
        /// <param name="capacity">The new capacity of the queue.</param>
        private void Resize(int capacity)
        {
            T[] temp = new T[capacity];
            for (int i = 0; i < N; i++)
            {
                temp[i] = q[(first + i) % q.Length];
            }
            q = temp;
            first = 0;
            last = N;
        }

        /// <summary>
        /// Adds the item to this queue.
        /// </summary>
        /// <param name="item">The item to add.</param>
        /// <exception cref="System.ArgumentNullException">Item cannot be null.</exception>
        public void Enqueue(T item)
        {
            if (item == null) { throw new ArgumentNullException(); }
            if (N == q.Length) Resize(q.Length * 2);
            q[last++] = item;
            if (last == q.Length) last = 0;
            N++;
        }

        /// <summary>
        /// Removes and returns the item on this queue that was least recently added.
        /// </summary>
        /// <returns>The item on this queue that was least recently added.</returns>
        /// <exception cref="System.ArgumentOutOfRangeException">Queue is empty.</exception>
        public T Dequeue()
        {
            if (IsEmpty()) { throw new ArgumentOutOfRangeException(); }
            T result = q[first];
            q[first] = default(T);
            N--;
            first++;
            if (first == q.Length) first = 0;
            if (N == q.Length / 4) Resize(q.Length / 2);
            return result;
        }

        /// <summary>
        /// Returns (but does not remove) the item least recently added to this queue.
        /// </summary>
        /// <returns>The item least recently added to this queue.</returns>
        /// <exception cref="System.ArgumentOutOfRangeException">Queue is empty.</exception>
        public T Peek()
        {
            if (IsEmpty()) { throw new ArgumentOutOfRangeException(); }
            return q[first];
        }

        /// <summary>
        /// Returns a string representation of this queue.
        /// </summary>
        /// <returns>The sequence of items in the queue in FIFO order, separated by spaces.</returns>
        public override string ToString()
        {
            return String.Join(" ", this);
        }

        /// <summary>
        /// Returns an enumerator that iterates through the items in this queue in FIFO order.
        /// </summary>
        /// <returns>
        /// A <see cref="T:System.Collections.Generic.IEnumerator`1" /> that can be used to iterate through the items in this queue in FIFO order.
        /// </returns>
        public IEnumerator<T> GetEnumerator()
        {
            return new Enumerator(this);
        }

        /// <summary>
        /// Returns an enumerator that iterates through the items in this queue in FIFO order.
        /// </summary>
        /// <returns>
        /// An <see cref="T:System.Collections.IEnumerator" /> object that can be used to iterate through the items in this queue in FIFO order.
        /// </returns>
        IEnumerator IEnumerable.GetEnumerator()
        {
            return new Enumerator(this);
        }

        private class Enumerator : IEnumerator<T>, IDisposable, IEnumerator
        {
            private int i = -1;
            private int N;
            private int first;
            private T[] q;

            public Enumerator(ResizingArrayQueue<T> queue)
            {
                first = queue.first;
                q = queue.q;
                N = queue.N;
            }

            public T Current
            {
                get { return q[(first + i) % q.Length]; }
            }

            public void Dispose()
            {
                i = N = first = 0;
                q = null;
            }

            object IEnumerator.Current
            {
                get { return q[(first + i) % q.Length]; }
            }

            public bool MoveNext()
            {
                return ++i < N;
            }

            public void Reset()
            {
                i = -1;
            }
        }
    }
}
